c7n/resources/s3.py

```diff
@@ -2203,3 +2203,87 @@ def process_chunk(self, batch, bucket):
             Bucket=bucket['Name'], Delete={'Objects': objects}).get('Deleted', ())
         if self.get_bucket_style(bucket) != 'versioned':
             return results
+
+
+@actions.register('configure-lifecycle')
+class Lifecycle(BucketActionBase):
+    """Action applies a lifecycle policy to versioned S3 buckets
+
+    The schema to supply to the rule follows the schema here: goo.gl/yULzNc
+
+    :example:
+
+        .. code-block: yaml
+
+            policies:
+              - name: s3-apply-lifecycle
+                resource: s3
+                actions:
+                  - type: configure-lifecycle
+                    rules:
+                      - ID: my-lifecycle-id
+                        Status: Enabled
+                        Prefix: foo/
+                        Transitions:
+                          - Days: 60
+                            StorageClass: GLACIER
+
+    """
+
+    schema = type_schema(
+        'configure-lifecycle',
+        **{
+            'rules': {
+                'type': 'array',
+                'required': True,
+                'items': {
+                    'type': 'object',
+                    'required': ['ID', 'Status'],
+                    'properties': {
+                        'ID': {'type': 'string'},
+                        'Status': {'type': 'string'},
+                    },
+                },
+            },
+        }
+    )
+
+    permissions = ('s3:GetLifecycleConfiguration', 's3:PutLifecycleConfiguration')
+
+    def get_lifecycle_configuration(self, s3, bucket):
+        # If no lifecycle exists AWS gives a 404.  Protect against that.
+        try:
+            resp = s3.get_bucket_lifecycle_configuration(Bucket=bucket)
+        except ClientError as e:
+            if e.response['Error']['Code'] == 'NoSuchLifecycleConfiguration':
+                resp = {}
+            else:
+                raise
+
+        return resp
+
+    def process(self, buckets):
+
+        for bucket in buckets:
+            s3 = bucket_client(local_session(self.manager.session_factory), bucket)
+
+            # put_bucket_lifecycle_configuration replaces any existing lifecycle
+            # with the supplied one.  So we need to fetch the existing one and
+            # merge the new one in appropriately.
+            resp = self.get_lifecycle_configuration(s3, bucket['Name'])
+            if 'Rules' not in resp:
+                config = self.data['rules']
+            else:
```

Comment:
this if/else construction can be simplified by returning {'Rules': []} in the case of not found instead of an empty dict.